python/cudf/cudf/tests/test_struct.py

```diff
@@ -371,3 +371,18 @@ def test_nested_struct_extract_host_scalars(data, idx, expected):
     series = cudf.Series(data)
 
     assert _nested_na_replace(series[idx]) == _nested_na_replace(expected)
+
+
+def test_struct_memory_usage():
+    s = cudf.Series([{"a": 1, "b": 10}, {"a": 2, "b": 20}, {"a": 3, "b": 30}])
+    df = s.struct.explode()
+
+    assert_eq(s.memory_usage(), df.memory_usage().sum())
+
+
+def test_struct_with_null_memory_usage():
+    s = cudf.Series(
+        [{"a": 1, "b": 10}, {"a": 2, "b": 20}, None, None, {"a": 3, "b": 30}]
+    )
+
+    assert s.memory_usage() == 272
```

Comment:
This makes some (implicit) assumptions on the storage format for python ints when going into cudf. Should we be explicit here?

Reply:
There's no easy way to achieve this thing with `.astype`, hence had to deconstruct and re-construct the struct series.

Reply:
`s = cudf.Series(..., dtype=cudf.StructDtype({"a": "int64", "b": "int64"}))`  ?

Reply:
It's not explicitly supported yet, and works only partially due to implicit support:
```
In [4]: s = cudf.Series([{"a": 1, "b": 10}, {"a": 2, "b": 20}, None, None, {"a": 3, "b": 30}], dtype=cudf.StructDtype({"a": "int64", "b": "float64"}))

In [5]: s
Out[5]: 
0    {'a': 1, 'b': 10.0}
1    {'a': 2, 'b': 20.0}
2                   None
3                   None
4    {'a': 3, 'b': 30.0}
dtype: struct

In [6]: s = cudf.Series([{"a": 1, "b": 10}, {"a": 2, "b": 20}, None, None, {"a": 3, "b": 30}], dtype=cudf.StructDtype({"a": "int64", "b": "str"}))
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
File /nvme/0/pgali/envs/cudfdev/lib/python3.9/site-packages/cudf/core/column/column.py:2031, in as_column(arbitrary, nan_as_null, dtype, length)
   2029 try:
   2030     data = as_column(
-> 2031         memoryview(arbitrary), dtype=dtype, nan_as_null=nan_as_null
   2032     )
   2033 except TypeError:

TypeError: memoryview: a bytes-like object is required, not 'list'

During handling of the above exception, another exception occurred:

ArrowTypeError                            Traceback (most recent call last)
File /nvme/0/pgali/envs/cudfdev/lib/python3.9/site-packages/cudf/core/column/column.py:2082, in as_column(arbitrary, nan_as_null, dtype, length)
   2079 elif isinstance(
   2080     dtype, cudf.StructDtype
   2081 ) and not isinstance(dtype, cudf.IntervalDtype):
-> 2082     data = pa.array(arbitrary, type=dtype.to_arrow())
   2083     return as_column(data, nan_as_null=nan_as_null)

File /nvme/0/pgali/envs/cudfdev/lib/python3.9/site-packages/pyarrow/array.pxi:317, in pyarrow.lib.array()

File /nvme/0/pgali/envs/cudfdev/lib/python3.9/site-packages/pyarrow/array.pxi:39, in pyarrow.lib._sequence_to_array()

File /nvme/0/pgali/envs/cudfdev/lib/python3.9/site-packages/pyarrow/error.pxi:144, in pyarrow.lib.pyarrow_internal_check_status()

File /nvme/0/pgali/envs/cudfdev/lib/python3.9/site-packages/pyarrow/error.pxi:123, in pyarrow.lib.check_status()

ArrowTypeError: Expected bytes, got a 'int' object

During handling of the above exception, another exception occurred:

TypeError                                 Traceback (most recent call last)
File /nvme/0/pgali/envs/cudfdev/lib/python3.9/site-packages/cudf/core/column/column.py:2192, in _construct_array(arbitrary, dtype)
   2191     dtype = dtype if dtype is None else cudf.dtype(dtype)
-> 2192     arbitrary = cupy.asarray(arbitrary, dtype=dtype)
   2193 except (TypeError, ValueError):

File /nvme/0/pgali/envs/cudfdev/lib/python3.9/site-packages/cupy/_creation/from_data.py:76, in asarray(a, dtype, order)
     50 """Converts an object to array.
     51 
     52 This is equivalent to ``array(a, dtype, copy=False)``.
   (...)
     74 
     75 """
---> 76 return _core.array(a, dtype, False, order)

File cupy/_core/core.pyx:2357, in cupy._core.core.array()

File cupy/_core/core.pyx:2381, in cupy._core.core.array()

File cupy/_core/core.pyx:2506, in cupy._core.core._array_default()

TypeError: Cannot interpret 'StructDtype({'a': dtype('int64'), 'b': dtype('O')})' as a data type

During handling of the above exception, another exception occurred:

TypeError                                 Traceback (most recent call last)
Cell In [6], line 1
----> 1 s = cudf.Series([{"a": 1, "b": 10}, {"a": 2, "b": 20}, None, None, {"a": 3, "b": 30}], dtype=cudf.StructDtype({"a": "int64", "b": "str"}))

File /nvme/0/pgali/envs/cudfdev/lib/python3.9/contextlib.py:79, in ContextDecorator.__call__.<locals>.inner(*args, **kwds)
     76 @wraps(func)
     77 def inner(*args, **kwds):
     78     with self._recreate_cm():
---> 79         return func(*args, **kwds)

File /nvme/0/pgali/envs/cudfdev/lib/python3.9/site-packages/cudf/core/series.py:536, in Series.__init__(self, data, index, dtype, name, nan_as_null)
    533         data = {}
    535 if not isinstance(data, ColumnBase):
--> 536     data = column.as_column(data, nan_as_null=nan_as_null, dtype=dtype)
    537 else:
    538     if dtype is not None:

File /nvme/0/pgali/envs/cudfdev/lib/python3.9/site-packages/cudf/core/column/column.py:2177, in as_column(arbitrary, nan_as_null, dtype, length)
   2172                 return cudf.core.column.ListColumn.from_sequences(
   2173                     arbitrary
   2174                 )
   2175             else:
   2176                 data = as_column(
-> 2177                     _construct_array(arbitrary, dtype),
   2178                     dtype=dtype,
   2179                     nan_as_null=nan_as_null,
   2180                 )
   2181 return data

File /nvme/0/pgali/envs/cudfdev/lib/python3.9/site-packages/cudf/core/column/column.py:2209, in _construct_array(arbitrary, dtype)
   2195     if (
   2196         dtype is None
   2197         and not cudf._lib.scalar._is_null_host_scalar(arbitrary)
   (...)
   2202         )
   2203     ):
   2204         native_dtype = "object"
   2205     arbitrary = np.asarray(
   2206         arbitrary,
   2207         dtype=native_dtype
   2208         if native_dtype is None
-> 2209         else np.dtype(native_dtype),
   2210     )
   2211 return arbitrary

TypeError: Cannot interpret 'StructDtype({'a': dtype('int64'), 'b': dtype('O')})' as a data type
```

Reply:
Aha, ok.